### ISSUE TITLE
Reading Midi Files

### ISSUE THREAD

**Issue by ha17** [stdtabs/phptabs#13]
Hi, I'm trying to update the names for some midi files as when I import them into my DAW, they show a standard name and it's tough to know what is what.

I'm doing this:

    <?php

    require_once 'vendor/autoload.php';

    use PhpTabs\PhpTabs;

    $file = $argv[1];
    $mid = new PhpTabs($file);
    echo $mid->getName() . PHP_EOL;

--

Each time the getName() is blank. If I use a gp* file, it seems to work fine (for example: https://gprotab.net/en/tabs/the-eagles/hotel-california works).

Here's an example of it not working for a Midi file: https://www.midiworld.com/download/3290 (Enter Sandman).

Thanks,
Hans

**Comment by landrok** [comment]
Hi Hans,

The method 

```php
echo $tab->getName() . PHP_EOL;
```
is a shortcut for 
```php
echo $tab->getSong()->getName() . PHP_EOL;
```
that returns a song name.

This metadata is clearly hold in GuitarPro formats but not as clearly in MIDI specs (META MESSAGES), as far as I know. It results that it's not often used.

A workaround may be to use the MIDI filename to set this meta:

```php
use PhpTabs\IOFactory;

$filename = 'folder/song_name.mid';
$tab = IOFactory::fromFile($filename);

// This is the part
if (is_null($tab->getName($filename))) {
    $tab->setName(str_replace('_', ' ', basename($filename)));
}

echo $tab->getName() . PHP_EOL;
```

Hope this helps,
Landrok

**Comment by ha17** [comment]
Thanks for your reply. I see the names in the binary data if I open in a text editor, so I assumed it was formatted correctly.  You may be correct that it isn't.   I'll see if I can use the file name and setName() method to properly set it. That could work.  I'll let you know after I try it.

**Comment by ha17** [comment]
Hi - this seemed to work, but when I save, something goes wrong and the midi file is empty. I've tried ->save(), ->save('filename.mid') and ->convert('mid'). Is there something more I need to do? I appreciate this script!

**Comment by landrok** [comment]
Hi @ha17 

Sorry for not having answered your questions before.

I've released PhpTabs 0.6.3 which should fix this problem. MIDI timing may be improved for complex tablatures, it's a work in progress.

Can you test it and reopen if needed ?
